refactor: Cohorts internals, use @property, add type hinting and expose it in docs

thatlittleboy opened the pull request:
## Overview

Description of the changes proposed in this pull request:
- Use `@property` for `Cohorts.cohorts` to enforce that the value passed in is always a dictionary of Explanation objects *before* the value is set. I believe this is the more pythonic way of enforcing this constraint, rather than doing it only in `__init__`. A side effect here of using the property descriptor is that users will no longer be able to `del` the cohorts attribute, but I guess that's not a "normal" thing to do..
- Added docstrings with examples to the Cohorts class. And made `Cohorts` a publicly documented feature (it's already in our public API being exposed in our `__init__.py`).
- <s>One benefit from doing this refactoring is that I spotted a bug in the definition of `Cohorts` (mypy, actually), in that the `Cohorts.__call__()` definition does not make sense. It calls `Explanation.__call__()` which does not exist. See: </s>

<img src="https://github.com/user-attachments/assets/8c6e161a-9ff0-4ed4-ba00-dae9139df4e4" width=300>

(minimized the image because it's not accurate, see follow up comment)

- EDIT: It turns out that `Cohorts.__call__` was actually included here to invoke methods on the individual Explanation objects, not for doing `Explanation.__call__` (which doesn't exist) directly. An example (also included in the newly added tests) would be the following:

```python
exp = shap.Explanation(
    values=rs.uniform(low=-1, high=1, size=e_size),
    data=rs.normal(loc=1, scale=3, size=e_size),
    feature_names=list("abcde"),
)
exp_neg = exp[exp[:, "a"].data < 0]
exp_pos = exp[exp[:, "a"].data >= 0]

ch = shap.Cohorts(col_a_neg=exp_neg, col_a_pos=exp_pos)

# ----------- here is the crux:
new_ch = ch.mean(axis=0)
# This does a `getattr(ch, "mean")` first on the Cohorts object, which results in a bunch of `Explanation.mean` bound methods;
# A Cohorts object is being returned by getattr, hence the subsequent `(axis=0)` is being invoked as a Cohorts.__call__
```

Thus, I've retained the `__call__` magic method on `Cohorts`, but storing the bound methods in a separate private variable, and these would be individually invoked during the `Cohorts.__call__`.

- Added tests to test all of the above mentioned intricacies.


## Checklist

- [x] All [pre-commit checks](https://pre-commit.com/#install) pass.
- [x] Unit tests added (if fixing a bug or adding a new feature)

thatlittleboy committed:
refactor: use property to enforce types in cohorts

thatlittleboy committed:
fix: remove __call__ since Explanation has no call

NOTE: this is because we had enforced `cohorts[k]` to always be an
Explanation upon property setting.
And Explanation has no `__call__` attribute (error thrown by mypy), so
it makes no sense for cohorts to have it as well.

thatlittleboy commented on shap/_explanation.py:
```diff
@@ -826,33 +826,39 @@ def _compute_shape(x) -> tuple[Union[int, None], ...]:
-    def __init__(self, **kwargs):
+    def __init__(self, **kwargs) -> None:
         self.cohorts = kwargs
-        for k in self.cohorts:
-            assert isinstance(
-                self.cohorts[k], Explanation
-            ), "All the arguments to a Cohorts set must be Explanation objects!"
```
This is now enforced in the property setter

thatlittleboy commented on shap/_explanation.py:
```diff
@@ -826,33 +826,39 @@ def _compute_shape(x) -> tuple[Union[int, None], ...]:
-            new_cohorts.cohorts[k] = self.cohorts[k].__call__(*args, **kwargs)
+        for k in self._cohorts:
+            new_cohorts.cohorts[k] = getattr(self._cohorts[k], name)
         return new_cohorts
```
`__call__()` is removed here, it makes no sense once we enforce that `self.cohorts[k]` must be an Explanation object.

thatlittleboy replied:
NOTE to self: The above claim is somewhat wrong. We still need to retain `__call__` here in order to invoke any methods that were retrieved during `__getattr__`.

E.g. `Cohorts(...).mean(axis=0)` would first run `__getattr__("mean")` and return a bound method `Explanation.mean` for all the Explanation objects inside the Cohorts. And return a Cohorts object.
Then the `(axis=0)` call would be executed on that returned Cohorts object, which is why we need `__call__`.

connortann replied:
Nice! To help out our future selves, I think it would help to include your explanation above as a comment in the `__call__` method, as it's quite subtle.

thatlittleboy replied:
done

connortann commented on shap/_explanation.py:
```diff
@@ -826,33 +826,39 @@ def _compute_shape(x) -> tuple[Union[int, None], ...]:
 
 class Cohorts:
-    def __init__(self, **kwargs):
+    def __init__(self, **kwargs) -> None:
```
This might be a good opportunity to add a class docstring to explain what a cohort represents and how the class is intended to be used.

We don't need to go overboard, but a few docstrings might help future devs get up to speed faster.

connortann replied:
PS, I think this `kwargs` could be type hinted as:

```suggestion
    def __init__(self, **kwargs: Explanation) -> None:
```

Using syntax explained here:
https://peps.python.org/pep-0484/#arbitrary-argument-lists-and-default-argument-values

thatlittleboy replied:
Thanks, added docstring with examples, please have a read and see if it makes sense.

connortann replied:
Looks great!

connortann commented on shap/_explanation.py:
```diff
@@ -826,33 +826,39 @@ def _compute_shape(x) -> tuple[Union[int, None], ...]:
+            raise TypeError(emsg)
+        if not all(isinstance(exp, Explanation) for exp in cval.values()):
+            emsg = "All the arguments to a Cohorts set must be Explanation objects!"
+            raise ValueError(emsg)
```
Should this be:
```suggestion
            raise TypeError(emsg)
```

thatlittleboy committed:
docs: Expose Cohorts as public API in the docs

thatlittleboy committed:
apply review suggestions

thatlittleboy committed:
fix: populate cohorts dict rather than modify one by one

This prevents the subtle bug where we can accidentally add in a non-Explanation object

thatlittleboy committed:
chore: add tests for Cohorts

thatlittleboy committed:
fix: add __call__ back

for invoking methods after they have been getattr-ed, e.g. cohort.mean(axis=0)

thatlittleboy committed:
fix: mypy complaints

connortann commented on shap/_explanation.py:
```diff
@@ -852,36 +856,99 @@ def _compute_shape(x) -> tuple[int | None, ...]:
+            raise TypeError(emsg)
+        if not all(isinstance(exp, Explanation) for exp in cval.values()):
+            emsg = "All the arguments to a Cohorts set must be Explanation objects!"
+            raise TypeError(emsg)
```
Nitpick: it might help to include the unexpected type in the exception message.

```suggestion
        for exp in cval.values():
            if not isinstance(exp, Explanation):
                emsg = f"Arguments to a Cohorts set must be Explanation objects, but found {type(exp)}"
                raise TypeError(emsg)
```

connortann commented on shap/_explanation.py:
```diff
@@ -852,36 +856,99 @@ def _compute_shape(x) -> tuple[int | None, ...]:
+            raise TypeError(emsg)
+
+        cast(dict[str, Explanation], cval)  # type narrowing for mypy
+        self._cohorts: dict[str, Explanation] = cval
```
I think the type narrowing applies to the _return value_ of `typing.cast`, so I think it is typically used like this:

```suggestion
        self._cohorts: dict[str, Explanation] = cast(dict[str, Explanation], cval)  # type narrowing for mypy
```

However, it might be that the assert statements above provide sufficient type narrowing, so maybe we don't need the `cast` call?

thatlittleboy replied:
Yeah, you're right. For some reason, I don't think mypy picked up that `cval` could only be a dict containing `Explanation`s when it was a generator comprehension (or I'm going crazy).

Anyway now it works after testing locally, so I've implemented your suggestion. Thanks~

connortann reviewed (approved):
Some minor suggestions but otherwise looks great. Nice one!

thatlittleboy committed:
include review suggestion

Co-authored-by: connortann <71127464+connortann@users.noreply.github.com>

thatlittleboy committed:
docs: add comment to docstring

thatlittleboy committed:
fix: remove cast after assert
